bolt/lib/Profile/DataAggregator.cpp

```diff
@@ -774,42 +774,86 @@ bool DataAggregator::doInterBranch(BinaryFunction *FromFunc,
 }
 
 bool DataAggregator::doBranch(uint64_t From, uint64_t To, uint64_t Count,
-                              uint64_t Mispreds) {
-  bool IsReturn = false;
-  auto handleAddress = [&](uint64_t &Addr, bool IsFrom) -> BinaryFunction * {
-    if (BinaryFunction *Func = getBinaryFunctionContainingAddress(Addr)) {
-      Addr -= Func->getAddress();
-      if (IsFrom) {
-        auto checkReturn = [&](auto MaybeInst) {
-          IsReturn = MaybeInst && BC->MIB->isReturn(*MaybeInst);
-        };
-        if (Func->hasInstructions())
-          checkReturn(Func->getInstructionAtOffset(Addr));
-        else
-          checkReturn(Func->disassembleInstructionAtOffset(Addr));
-      }
+                              uint64_t Mispreds, bool IsPreagg) {
+  // Returns whether \p Offset in \p Func contains a return instruction.
+  auto checkReturn = [&](const BinaryFunction &Func, const uint64_t Offset) {
+    auto isReturn = [&](auto MI) { return MI && BC->MIB->isReturn(*MI); };
+    return Func.hasInstructions()
+               ? isReturn(Func.getInstructionAtOffset(Offset))
+               : isReturn(Func.disassembleInstructionAtOffset(Offset));
+  };
 
-      if (BAT)
-        Addr = BAT->translate(Func->getAddress(), Addr, IsFrom);
+  // Returns whether \p Offset in \p Func corresponds to a call continuation
+  // fallthrough block.
+  auto checkCallCont = [&](BinaryFunction &Func, const uint64_t Offset) {
+    // Note the use of MCInstrAnalysis: no call continuation for a tail call.
+    auto isCall = [&](auto MI) { return MI && BC->MIA->isCall(*MI); };
+
+    // No call continuation at a function start.
+    if (!Offset)
+      return false;
+
+    // FIXME: support BAT case where the function might be in empty state
```

Comment:
Are you working on a follow-up PR or what's the plan otherwise?

Reply:
The current plan wrt BAT mode:
0) The current version has some support (non-PLT calls in simple functions).
1) Directly check that branch source is a return instruction. This gets us to covering all non-PLT calls. Will include in this PR.
2) PLT calls in non-simple functions: either revive #90806, or keep call offsets in BinaryFunction (we have that for taken branches). What's your opinion on this?

Reply:
If we have an external branch, and the destination is neither a landing pad nor an entry point, *and* the instruction preceding the destination is a call, then it strongly smells like a return branch to me and I don't see a reason for any further checks, like for a return instruction at the branch source.

Reply:
We can't check if the destination is LP/entry or check the preceding instruction unless the function is in disassembled/CFG state. And non-simple functions in BAT mode are not. 
We can check if the source is a return instruction if it's local.

Reply:
Perhaps we should add secondary entry point and EH info to BAT then. It's important enough from CFG point of view.

Reply:
We can do that, secondary entry points are already in BAT. The missing piece is checking if the preceding instruction of a destination is a call. Let's discuss that offline. 

I'll start working on a follow-up.

Reply:
Follow-up BAT support: https://github.com/llvm/llvm-project/pull/114602 and #115334